### DISCUSSION TITLE
Ledger integration feedback

### DISCUSSION THREAD

**Discussion by kyranjamie** [blockstack/stacks-wallet-web#2275] [General]
Opening this discussion to collect feedback on the Ledger integration 📒

It's still under active development, but community testers welcome 🐛

[Download the beta version of the Hiro Wallet supporting Ledger →](https://github.com/hirosystems/stacks-wallet-web/pull/2242)

Here's a demo of the proof of concept running through the happy flow

https://user-images.githubusercontent.com/1618764/156131221-7e41f508-3c3a-4c33-a4d0-f89b2628c9be.mp4

**Comment by kyranjamie** [comment]
Here are my initial questions/ideas to improve the onboarding from the current design spec.

<details>
<summary>1. Home screen prominence of Ledger action</summary>
<br />

![image](https://user-images.githubusercontent.com/1618764/156125031-98269559-c4c0-41ab-bd6b-0644ef39cfec.png)

I wonder whether we should add a slightly more prominent _Connect your Ledger_ CTA. We should be encouraging this, and creating a software account is the more heavy action. Should we clarify that _Create Stacks account_ makes a software wallet? Should they have a second chance to choose Ledger on the next step?

</details>

<details>
<summary>2. Set user expectation that WebUSB pop up will open</summary>
<br />

![image](https://user-images.githubusercontent.com/1618764/156123287-d3ef86c3-5877-4950-96a0-f6ee2be1672c.png)

Some text/imagery that sets the user's expectation that this window will open would improve this step. Connecting a USB device isn't a super common action, so some guidance would help novice users.

Maybe we show a message after clicking the _Connect_ button saying something along the lines of "Find your Ledger device and press connect in the pop up", and after a short delay, trigger the action?

**Note**: The _Nano S_ in this window (below) will show when the Ledger is on the _Select your app_ screen, as well as when the Stacks app is open. But the _Stacks Ledger app_ must be open for onboarding to work. If we can get the user to have this app open/ready _before_ pressing connect, it'll be a much better ux.

![image](https://user-images.githubusercontent.com/1618764/156123873-e6e39347-0f21-4df0-b443-bf8c8417f330.png)

</details>

<details>
<summary>3. Alternate copy while <em>Looking for device</em></summary>
<br />

![image](https://user-images.githubusercontent.com/1618764/156124737-2a3f2a6d-8a93-427f-91a9-e3edad848777.png)

By the time we get this step, we've already found the device. It takes milliseconds for us to establish whether or not the device is connected, or if there's a connection error. There is not "Looking" time. 

During this time, we pull _n_ public keys from the device. 5 public keys, the current default, takes ~10 seconds.

</details>

<details>
<summary>4. Common error handling</summary>
<br />
Common, easily-fixed error cases, such as the device being locked, or the app not being open should be handled in the least disruptive manner possible. Navigating to a catch-all error screen is a heavy-handed error flow in these cases (useful, but for transport errors, such as if you unplug the device while actions are in-flight). 
<br /><br />

We should consider how we handle these inline. Should we use the UI below?

![image](https://user-images.githubusercontent.com/1618764/156126709-765d7e2e-120f-48cd-a16a-c408c0a6f58f.png)

</details>

cc/ @fbwoolf @landitus

**Comment by markmhendrickson** [comment]
I've been discussing a holistic re-look at the onboarding approach with @landitus recently, and I imagine much of what we can do to improve Ledger-specific onboarding can be incorporated into that effort (and unlinked from this initial Ledger release).

As such, I would suggest we focus on base-level design needs here that are missing, such as any common errors that aren't clear already for this initial release.

@kyranjamie can you provide a list of all possible errors and an indication of when and why they arise?

**Comment by fbwoolf** [comment]
Notes from our coordination meeting this morning:

- Remove Lock from settings
- Remove Create an Account from settings
- We will automatically pull in 3-5 accounts for the user until we design the flow to create an account
- Identify in UI that the user is connected to a hardware wallet
- Should the tx signing flow stay in the drawer/modal rather than move to full page?

**Comment by landitus** [comment]
Thanks for putting this together @kyranjamie. I agree with all your remarks about improvements to the UX 💯!!
I will focus on the missing states and errors in the short term, to have them for the release.

**Comment by landitus** [comment]
Some design updates for the Ledger Transaction Signing. I've mocked the happy path and also the error states that can come up during the transaction. 

### User flows
<img width="4100" alt="Ledger - Transaction signing flow" src="https://user-images.githubusercontent.com/239215/159771945-1711b992-99d6-4509-a12a-0b245760063c.png">

### Prototype
Check this [Figma](https://www.figma.com/proto/VQPOPkkrfVcf0BEd9kz3HZ/%232275---Ledger%3A-Onboarding-%2B-Transaction-signing?page-id=505%3A14264&node-id=505%3A14336&viewport=328%2C48%2C0.13&scaling=scale-down&starting-point-node-id=505%3A14336&show-proto-sidebar=1) for all the different flows.

In the prototype, it assumes the Ledger is already plugged in and has been paired once to the Chrome browser. We may need to mock a state where the Ledger is not even plugged in.
